### PR TITLE
feat(javaagent): support custom instrumentations (Issue #287)

### PR DESCRIPTION
**Pull request by Pittu-Sharma**
[PR Description]
Motivation
Currently, the OpenTelemetry Java Agent registry defines core non-library features (such as Methods, JMX metrics, and Annotations) within a custom array in the instrumentation.yaml inventory. The Ecosystem Explorer was previously only processing the libraries list, causing these essential instrumentations to be missing from the web application.

This PR addresses Issue #287 by enabling full support for custom instrumentations across the entire data pipeline—from ingestion and transformation to frontend rendering.

Changes
Backend (ecosystem-automation/explorer-db-builder)

Transformation Logic: Refactored instrumentation_transformer.py to apply schema upgrades (v0.1 through v0.5) to both library and custom instrumentation lists.
Database Manifests: Updated database_writer.py to include a custom_instrumentations key in the generated version manifests, mapping IDs to content hashes.
Metadata Enrichment: Enhanced the main.py orchestrator to perform metadata backfilling for custom instrumentations, ensuring descriptions and display names are preserved across versions.
Frontend (ecosystem-explorer)

API & Typing: Extended the VersionManifest and InstrumentationData types to handle multiple instrumentation maps.
Segmented UI: Updated JavaInstrumentationListPage to group instrumentations by source. Introduced a new "Custom Instrumentations" section with descriptive subtitles to clearly delineate them from standard libraries.
Accessibility: Implemented semantic HTML (<h2>) and verified aria-pressed states for filters to ensure compliance with the project's accessibility checklist.
Verification
Backend Tests: All 174 Python tests passed (uv run pytest).
Frontend Tests: All 498 Vitest tests passed (bun run test).
Local Build: Successfully rebuilt the database and verified that custom instrumentations (like Methods and JMX) are correctly ingested and rendered.
Compliance: Passed all quality gates including ruff linting, markdownlint, eslint, and copyright header checks.

### CODE REVIEW

File: ecosystem-explorer/src/types/javaagent.ts [hunk 1]
```diff
@@ -43,6 +44,7 @@ export interface InstrumentationData {
   javaagent_target_versions?: string[];
   configurations?: Configuration[];
   telemetry?: Telemetry[];
+  _isCustom?: boolean;
```
Comment (jaydeluca) [inline ecosystem-explorer/src/types/javaagent.ts]: Can we change this to match the casing of all the other fields?

```suggestion
  _is_custom?: boolean;
```

File: ecosystem-explorer/src/features/java-agent/java-instrumentation-list-page.tsx [hunk 2]
```diff
@@ -192,15 +203,44 @@ export function JavaInstrumentationListPage() {
...
+                  {libraryGroups.map((group) => (
+                    <InstrumentationGroupCard
+                      key={group.displayName}
+                      group={group}
+                      activeFilters={filters}
+                      version={resolvedVersion}
+                    />
+                  ))}
+                </div>
+              </div>
```
Comment (jaydeluca) [inline ecosystem-explorer/src/features/java-agent/java-instrumentation-list-page.tsx:217]: We have an extra `<div>` in here we can remove

```suggestion
            {libraryGroups.length > 0 && (
              <div className="grid gap-6 md:grid-cols-2">
                {libraryGroups.map((group) => (
                  <InstrumentationGroupCard
                    key={group.displayName}
                    group={group}
                    activeFilters={filters}
                    version={resolvedVersion}
                  />
                ))}
              </div>
```

File: ecosystem-automation/explorer-db-builder/src/explorer_db_builder/main.py [hunk 3]
```diff
@@ -134,12 +138,18 @@ def run_javaagent_builder(
...
 
+        backfilled_inventories = backfill_metadata(
+            versions,
+            lambda v: backfilled_libraries.get(v) or inventory_manager.load_versioned_inventory(v),
+            item_key="custom",
+        )
+
         for version in versions:
             inventory = backfilled_inventories.get(version)
             process_version(version, inventory_manager, db_writer, inventory=inventory)
```
Comment (jaydeluca) [inline ecosystem-automation/explorer-db-builder/src/explorer_db_builder/main.py:154]: I think it's a little hard to reason about the `lambda v: backfilled_libraries.get(v) or inventory_manager.load_versioned_inventory(v)` part, what if instead we just did two separate backfills and then combine them in the version loop, i think it's a little easier to follow

```suggestion
        backfilled_libraries = backfill_metadata(
            versions,
            inventory_manager.load_versioned_inventory,
            item_key="libraries",
        )
        backfilled_customs = backfill_metadata(
            versions,
            inventory_manager.load_versioned_inventory,
            item_key="custom",
        )

        for version in versions:
            inventory = backfilled_libraries[version].copy()
            inventory["custom"] = backfilled_customs[version].get("custom", [])
            process_version(version, inventory_manager, db_writer, inventory=inventory)
```

File: ecosystem-explorer/src/features/java-agent/java-instrumentation-list-page.tsx [hunk 4]
```diff
@@ -101,9 +101,21 @@ export function JavaInstrumentationListPage() {
...
 
-  const groupedInstrumentations = useMemo(
-    () => groupInstrumentationsByDisplayName(filteredInstrumentations),
-    [filteredInstrumentations]
+  const { libraryInstrumentations, customInstrumentations } = useMemo(() => {
+    return {
+      libraryInstrumentations: filteredInstrumentations.filter((i) => !i._isCustom),
+      customInstrumentations: filteredInstrumentations.filter((i) => i._isCustom),
+    };
+  }, [filteredInstrumentations]);
```
[the remark this answers is not in the document]
Comment (jaydeluca) [reply ecosystem-explorer/src/features/java-agent/java-instrumentation-list-page.tsx:109]: you can ignore this for now, we'll handle it in a followup

File: ecosystem-automation/explorer-db-builder/tests/test_database_writer.py [hunk 5]
```diff
@@ -244,14 +228,12 @@ def test_write_version_index_creates_directory(self, db_writer, temp_db_dir):
         assert versions_dir.is_dir()
 
     def test_write_version_index_empty_map(self, db_writer):
-        """Empty library map raises ValueError."""
         version = Version("1.0.0")
 
-        with pytest.raises(ValueError, match="Library map cannot be empty"):
+        with pytest.raises(ValueError, match="Library map and custom map cannot both be empty"):
             db_writer.write_version_index(version, {})
```
[the remark this answers is not in the document]
Comment (jaydeluca) [reply ecosystem-automation/explorer-db-builder/tests/test_database_writer.py:235]: i agree with adding a new test for writing both `instrumentation` and `custom_instrumentations`, but "allowing an empty library_map when custom_map is non-empty" is not a realistic scenario in our case, so you can ignore that one.

File: ecosystem-automation/explorer-db-builder/tests/test_main.py [hunk 6]
```diff
@@ -94,54 +90,53 @@ def test_get_release_versions_filters_prereleases(self, mock_inventory_manager):
...
+        # write_libraries will be called twice (libraries, custom)
+        mock_db_writer.write_libraries.side_effect = [library_map, custom_map]
 
         process_version(version, mock_inventory_manager, mock_db_writer)
 
         mock_inventory_manager.load_versioned_inventory.assert_called_once_with(version)
-        mock_db_writer.write_libraries.assert_called_once_with(inventory_data["libraries"])
-        mock_db_writer.write_version_index.assert_called_once_with(version, library_map)
+        assert mock_db_writer.write_libraries.call_count == 2
+        mock_db_writer.write_version_index.assert_called_once_with(version, library_map, custom_map)
```
[the remark this answers is not in the document]
Comment (jaydeluca) [reply ecosystem-automation/explorer-db-builder/tests/test_main.py:113]: this is unecessary, you can ignore

**Review by jaydeluca — COMMENTED** [review]
really great job! I've left some comments, and I've reviewed the copilot suggestions, I think they are valid with the exception of the ones I commented on

**Comment by Pittu-Sharma** [comment]
@jaydeluca ,
I have made the changes  and updated the PR again as my checks are failing. Please review it and if any changes are required please let me know . I will work on that .

File: .github/workflows/screenshots.yml [hunk 7]
```diff
@@ -23,19 +23,35 @@ jobs:
...
-        uses: oven-sh/setup-bun@0c5077e51419868618aeaa5fe8019c62421857d6 # v2.2.0
+        uses: oven-sh/setup-bun@4bc823d497938332a49074c35903af2f23513361 # v2.0.1
         with:
-          bun-version: "1.3.12"
+          bun-version: "1.1.34"
 
       - name: Install dependencies
         run: bun install --frozen-lockfile
 
+      - name: Install uv
```
Comment (jaydeluca) [inline .github/workflows/screenshots.yml]: i dont think we need to do any changes to this file, we shouldn't need to regenerate the db for the screenshots

File: .github/workflows/build-and-test.yml [hunk 8]
```diff
@@ -23,28 +23,32 @@ jobs:
   check-copyright-headers:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
       - name: Install uv
-        uses: astral-sh/setup-uv@cec208311dfd045dd5311c1add060b2062131d57 # v8.0.0
+        uses: astral-sh/setup-uv@1ed1432da45330752f20542383139370830bb62c # v5.1.0
```
Comment (jaydeluca) [inline .github/workflows/build-and-test.yml]: looks like you have some merge conflicts that need to be addressed

**Review by jaydeluca — COMMENTED** [review]
looks good! just one last comment and then you just have some merge conflicts to address and this should be good to go

**Comment by Pittu-Sharma** [comment]
> looks good! just one last comment and then you just have some merge conflicts to address and this should be good to go

I have resolved the merge conflict issue. Is anything still left or am I got to go with this issue now .
Please give your suggestion.

File: .github/workflows/screenshots.yml [hunk 9]
```diff
@@ -36,6 +36,22 @@ jobs:
...
+        uses: astral-sh/setup-uv@08807647e7069bb48b6ef5acd8ec9567f424441b # v8.1.0
```
Comment (jaydeluca) [inline .github/workflows/screenshots.yml]: can you just revert the changes to this file? we shouldn't need to touch this workflow

Comment (Pittu-Sharma) [reply .github/workflows/screenshots.yml]: I have reverted the changes to this file.

**Review by jaydeluca — APPROVED** [review]
thanks @Pittu-Sharma, great contribution !
